### PR TITLE
ad: bug fixes on error handling and sp creating from external apps

### LINKED ISSUE #6439: az cli unable to create service principal for multi tenant applications
AD service prinical creation fails from the az cli when the application is a multi tenant application in a different tenant. It works if the application is in the same tenant.

az ad sp create --id <redacted>
Resource '<redacted>' does not exist or one of its queried reference-property objects are not present.
Traceback (most recent call last):
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\knack\cli.py", line 197, in invoke
    cmd_result = self.invocation.execute(args)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 347, in execute
    six.reraise(*sys.exc_info())
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\six.py", line 693, in reraise
    raise value
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 319, in execute
    result = cmd(params)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 180, in __call__
    return super(AzCliCommand, self).__call__(*args, **kwargs)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\knack\commands.py", line 109, in __call__
    return self.handler(*args, **kwargs)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\__init__.py", line 420, in default_command_handler
    result = op(**command_args)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\command_modules\role\custom.py", line 697, in create_service_principal
    return _create_service_principal(cmd.cli_ctx, identifier)
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\command_modules\role\custom.py", line 711, in _create_service_principal
    result = [client.applications.get(identifier)]
  File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\graphrbac\operations\applications_operations.py", line 265, in get
    raise models.GraphErrorException(self._deserialize, response)
azure.graphrbac.models.graph_error.GraphErrorException: Resource 'redactted' does not exist or one of its queried reference-property objects are not present.

**To Reproduce**
CLI version is 2.0.30
az ad sp create --id <redacted>

**Expected behavior**
The service pricipal should get created

**Environment summary**
Windows server, installed with the msi
CLI version is 2.0.30

**Additional context**

### PR DESCRIPTION
**Pull request by yugangw-msft**
Fix #6439

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/commands.py [hunk 1]
```diff
@@ -77,42 +85,43 @@ def load_command_table(self, _):
         g.custom_command('create', 'create_role_assignment')
         g.custom_command('list-changelogs', 'list_role_assignment_change_logs')
 
-    with self.command_group('ad app', client_factory=get_graph_client_applications, resource_type=PROFILE_TYPE, min_api='2017-03-10') as g:
+    with self.command_group('ad app', client_factory=get_graph_client_applications, resource_type=PROFILE_TYPE,
+                            min_api='2017-03-10', exception_handler=graph_err_handler) as g:
         g.custom_command('create', 'create_application')
         g.custom_command('delete', 'delete_application')
         g.custom_command('list', 'list_apps')
-        g.custom_command('show', 'show_application', exception_handler=empty_on_404)
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/commands.py]: `empty_on_404` never works as it only captures CloudError, but all graph commands raises `GraphErrorException`

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -705,19 +705,20 @@ def create_service_principal(cmd, identifier):
...
-            result = [client.applications.get(identifier)]
-        app_id = result[0].app_id
-    else:
-        app_id = identifier
+        try:
+            if not result:  # assume we get an object id
+                result = [client.applications.get(identifier)]
+            app_id = result[0].app_id
+        except GraphErrorException:
+            pass  # fallback to appid (maybe from an external tenant?)
```
Comment (colemickens) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: How does this work? I would have assumed that the SP would have to somehow more directly reference the App, if it's not in the sub? Does AD just "find" the App ID at runtime, even though the CLI doesn't see it?

Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: That is correct. AAD would be able to find it as long as you have configured that APP as `multi-tenant` in the original tenant.
CLI used to error out when can't find it in the current tenant, but now let service end do the last try. If the APP-ID is not right, users still get an error.

Comment (colemickens) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Cool. Thanks for the details!

**Comment by yugangw-msft** [comment]
//cc: @troydai

**Pull request merged**